test/unit/sql/insert.test.js

```diff
@@ -66,6 +66,68 @@ describe(Support.getTestDialectTeaser('SQL'), () => {
     });
   });
 
+  if (current.dialect.supports.inserts.onConflictWhere){
+    it('adds conflictWhere clause to generated queries', () => {
+      const User = Support.sequelize.define(
+        'user',
+        {
+          username: {
+            type: DataTypes.STRING,
+            field: 'user_name',
+            primaryKey: true,
+          },
+          password: {
+            type: DataTypes.STRING,
+            field: 'pass_word',
+          },
+          createdAt: {
+            type: DataTypes.DATE,
+            field: 'created_at',
+          },
+          updatedAt: {
+            type: DataTypes.DATE,
+            field: 'updated_at',
+          },
+        },
+        {
+          timestamps: true,
+        },
+      );
+
+      const upsertKeys = User.primaryKeyAttributes.map(
+        attr => User.rawAttributes[attr].field || attr,
+      );
+
+      expectsql(
+        sql.insertQuery(
+          User.tableName,
+          { user_name: 'testuser', pass_word: '12345' },
+          User.fieldRawAttributesMap,
+          {
+            updateOnDuplicate: ['user_name', 'pass_word', 'updated_at'],
+            conflictWhere: {
+              user_name: 'test where value',
+            },
+            upsertKeys,
+          },
+        ),
+        {
+          default:
+            'INSERT INTO `users` (`user_name`,`pass_word`) VALUES ($1,$2);',
+          postgres:
+            'INSERT INTO "users" ("user_name","pass_word") VALUES ($1,$2) ON CONFLICT ("user_name") WHERE "user_name" = \'test where value\' DO UPDATE SET "user_name"=EXCLUDED."user_name","pass_word"=EXCLUDED."pass_word","updated_at"=EXCLUDED."updated_at";',
+          mssql: 'INSERT INTO [users] ([user_name],[pass_word]) VALUES ($1,$2);',
```

Comment:
IMO if `conflictWhere` is not supported by the dialect, the query generator should throw an error.

It doesn't match how things were done before (i.e. 'updateOnDuplicate' is ignored if unsupported) but I think we should slowly make Sequelize safe by default (not generating part of a query is unsafe as devs may think their query works).

Discussable of course :)